db/hints/sync_point_service.cc

```diff
@@ -0,0 +1,442 @@
+            return ret;
+        });
+    });
+}
+
+future<> sync_point_service::unregister_rpc_verbs() {
+    auto& ms = _messaging;
+
+    return when_all(
+        ms.unregister_hint_sync_point_create(),
+        ms.unregister_hint_sync_point_check()
+    ).discard_result();
+}
+
+future<> sync_point_service::coordinate_hint_waiting(
+        abort_source& as,
+        sync_point_id id,
+        std::vector<gms::inet_address> source_nodes,
+        std::vector<gms::inet_address> target_nodes) {
+    
+    assert(this_shard_id() == 0);
+
+    // We want to abort waiting for all follower sync points either if waiting
+    // for one of the follower failed, or the operation was aborted.
+    // This abort_source will be triggered if either of those situations happen.
+    abort_source combined_as;
+    auto trigger_on_operation_abort_sub = as.subscribe([&combined_as, id] () noexcept {
+        service_logger.debug("Task ({} | coordinator): requested abort from above", id);
+        if (!combined_as.abort_requested()) {
+            combined_as.request_abort();
+        }
+    });
+
+    co_await parallel_for_each(source_nodes, [&ms_ = _messaging, &combined_as_ = combined_as, id_ = id, &target_nodes_ = target_nodes] (gms::inet_address follower) -> future<> {
+        auto& ms = ms_;
+        auto& combined_as = combined_as_;
+        const auto id = id_;
+        auto& target_nodes = target_nodes_;
+
+        service_logger.debug("Task ({} | coordinator): waiting for hints to be replayed towards {}", id, follower);
+
+        try {
+            // Create a hint sync point
+            sync_point_create_request request;
+            request.sync_point_id = id;
+            request.target_endpoints = target_nodes;
+            co_await ms.send_hint_sync_point_create(netw::msg_addr{ follower, 0 }, clock_type::time_point::max(), std::move(request));
+
+            // Wait in a loop until the sync point resolves
+            while (true) {
+                sync_point_check_request request;
+                request.sync_point_id = id;
+                auto resp = co_await ms.send_hint_sync_point_check(netw::msg_addr{ follower, 0 }, clock_type::time_point::max(), std::move(request));
+
+                if (resp.expired) {
+                    // Hints were successfully replayed - return
+                    break;
+                }
+
+                // Hints are still being replayed on the follower
+                // Wait a little before checking again
+                // If abort_source was engaged, this will throw - that's what we want
+                co_await sleep_abortable(follower_status_check_interval, combined_as);
+            }
+        } catch (...) {
+            // TODO: Print error?
+            service_logger.debug("Task ({} | coordinator): got an exception: {}", id, std::current_exception());
+            if (!combined_as.abort_requested()) {
+                combined_as.request_abort();
+            }
+            throw;
```

Comment:
What if a node is restarting?

Reply:
Then the whole waiting operation will fail and report a "FAILED" status.

Reply:
And already-created sync points will be garbage collected by timeouts?

Reply:
Follower sync points are GC'd 10 seconds after the last time they were queried.

Coordinator queries all follower sync points for this operation every second. If one node becomes unreachable, the coordinator will report an error and stop querying followers. After 10 seconds follower sync points should be GC'd.